plugins/experimental/log_requests/log_requests.cc

```diff
@@ -0,0 +1,352 @@
+{
+  if (version == TS_HTTP_VERSION(1, 0))
+    return "HTTP/1.0";
+  else if (version == TS_HTTP_VERSION(1, 1))
+    return "HTTP/1.1";
+  else if (version == TS_HTTP_VERSION(1, 2))
+    return "HTTP/1.2";
+  else if (version == TS_HTTP_VERSION(2, 0))
+    return "HTTP/2.0";
+  else {
+    return "(Unknown HTTP version)";
+  }
+}
+
+static void
+log_request_line(TSMBuffer bufp, TSMLoc loc, std::string output_header)
+{
+  int method_len;
+  int url_len;
+  const char *method;
+  const char *url;
+  TSMLoc url_loc;
+
+  // parse method
+  method = TSHttpHdrMethodGet(bufp, loc, &method_len);
+
+  // parse version
+  const std::string version = convert_http_version(TSHttpHdrVersionGet(bufp, loc));
+
+  // parse request line URL
+  TSHttpHdrUrlGet(bufp, loc, &url_loc);
+  url = TSUrlStringGet(bufp, url_loc, &url_len);
+
+  // get rid of the preceeding http:// on the request URI
+  for (int i = 0; i < 7; ++i)
+    url++;
+  url_len -= 7;
+
+  TSError(B_PLUGIN_NAME " [%s] request line is:\n%.*s %.*s %s\n", output_header.c_str(), method_len, method, url_len, url,
+          version.c_str());
+}
+
+static void
+log_response_status_line(TSMBuffer bufp, TSMLoc loc, std::string output_header)
+{
+  TSHttpStatus status_code;
+  const char *explanation;
+  int explanation_len;
+
+  // parse version
+  const std::string version = convert_http_version(TSHttpHdrVersionGet(bufp, loc));
+
+  // parse status code
+  status_code = TSHttpHdrStatusGet(bufp, loc);
+
+  // get explanation
+  explanation = TSHttpHdrReasonGet(bufp, loc, &explanation_len);
+
+  TSError(B_PLUGIN_NAME " [%s] response status line is:\n%s %d %.*s\n", output_header.c_str(), version.c_str(), status_code,
+          explanation_len, explanation);
+}
+
+static void
+log_headers(TSMBuffer bufp, TSMLoc loc, std::string output_header)
+{
+  TSIOBuffer output_buffer;
+  TSIOBufferReader reader;
+  TSIOBufferBlock block;
+  const char *block_start;
+  int64_t block_avail;
+
+  output_buffer = TSIOBufferCreate();
+  reader        = TSIOBufferReaderAlloc(output_buffer);
+
+  // This will print just MIMEFields and not the http request line
+  TSMimeHdrPrint(bufp, loc, output_buffer);
+
+  // We need to loop over all the buffer blocks, there can be more than 1
+  block = TSIOBufferReaderStart(reader);
+  do {
+    block_start = TSIOBufferBlockReadStart(block, reader, &block_avail);
+    if (block_avail > 0) {
+      TSError(B_PLUGIN_NAME " Headers are:\n%.*s", static_cast<int>(block_avail), block_start);
+    }
+    TSIOBufferReaderConsume(reader, block_avail);
+    block = TSIOBufferReaderStart(reader);
+  } while (block && block_avail != 0);
+
+  // Free up the TSIOBuffer that we used to print out the header
+  TSIOBufferReaderFree(reader);
+  TSIOBufferDestroy(output_buffer);
+}
+
+/*
+ * TODO: There are resource leaks if calls to TSHttpTxnClientReqGet() & TSHttpTxnClientRespGet()
+ * fail. Likewise for the ...ServerReqGet() & ...ServerRespGet() calls. Need to do work with
+ * some goto's to fix.
```

Comment:
I'm unsure how to use a smart pointer w/ a stack allocated pointer like `TSMBuffer`. I think I found another workaround anyways.